compiler/verification/z3encoding.ml

```diff
@@ -0,0 +1,632 @@
+      | Gte KInt | Gte KMoney | Gte KDate -> (ctx, Arithmetic.mk_ge ctx.ctx_z3 e1 e2)
+      | Gte _ ->
+          failwith
+            "[Z3 encoding] application of non-integer or money binary operator Gte not supported"
+      | Eq -> (ctx, Boolean.mk_eq ctx.ctx_z3 e1 e2)
+      | Neq -> (ctx, Boolean.mk_not ctx.ctx_z3 (Boolean.mk_eq ctx.ctx_z3 e1 e2))
+      | Map -> failwith "[Z3 encoding] application of binary operator Map not supported"
+      | Concat -> failwith "[Z3 encoding] application of binary operator Concat not supported"
+      | Filter -> failwith "[Z3 encoding] application of binary operator Filter not supported")
+  | Unop uop -> (
+      let ctx, e1 =
+        match args with
+        | [ e1 ] -> translate_expr ctx e1
+        | _ ->
+            failwith
+              (Format.asprintf "[Z3 encoding] Ill-formed unary operator application: %a"
+                 (Print.format_expr ctx.ctx_decl)
+                 (EApp ((EOp op, Pos.no_pos), args), Pos.no_pos))
+      in
+
+      match uop with
+      | Not -> (ctx, Boolean.mk_not ctx.ctx_z3 e1)
+      | Minus _ -> failwith "[Z3 encoding] application of unary operator Minus not supported"
+      (* Omitting the log from the VC *)
+      | Log _ -> (ctx, e1)
+      | Length -> failwith "[Z3 encoding] application of unary operator Length not supported"
+      | IntToRat -> failwith "[Z3 encoding] application of unary operator IntToRat not supported"
+      | GetDay -> failwith "[Z3 encoding] application of unary operator GetDay not supported"
+      | GetMonth -> failwith "[Z3 encoding] application of unary operator GetMonth not supported"
+      | GetYear -> failwith "[Z3 encoding] application of unary operator GetYear not supported")
+
+(** [translate_expr] translate the expression [vc] to its corresponding Z3 expression **)
+and translate_expr (ctx : context) (vc : expr Pos.marked) : context * Expr.expr =
+  let translate_match_arm (head : Expr.expr) (ctx : context)
+      (e : expr Pos.marked * FuncDecl.func_decl list) : context * Expr.expr =
+    let e, accessors = e in
+    match Pos.unmark e with
+    | EAbs (e, _) ->
+        (* Create a fresh Catala variable to substitue and obtain the body *)
+        let fresh_v = Var.make ("arm!tmp", Pos.no_pos) in
+        let fresh_e = EVar (fresh_v, Pos.no_pos) in
+
+        (* Invariant: Catala enums always have exactly one argument *)
+        let accessor = List.hd accessors in
+        let proj = Expr.mk_app ctx.ctx_z3 accessor [ head ] in
+        (* The fresh variable should be substituted by a projection into the enum in the body, we
+           add this to the context *)
+        let ctx = add_z3matchsubst fresh_v proj ctx in
+
+        let body = Bindlib.msubst (Pos.unmark e) [| fresh_e |] in
+        translate_expr ctx body
+    (* TODO: Denis, Is this true for constructors with no payload? *)
+    | _ -> failwith "[Z3 encoding] : Arms branches inside VCs should be lambdas"
```

Comment:
Right now all constructors have a payload. The "no payload" case is encoded with a unit payload and yes there is always a lambda.